File: archived/src/erlang/ch01/1.1/l0/src/bin_to_dec.erl
-module(bin_to_dec).
-export([convert/1]).

%% Convert a binary string like "1011" to integer.
convert(Str) when is_list(Str) ->
    lists:foldl(fun(Char, Acc) ->
        case Char of
            $0 -> Acc * 2;
            $1 -> Acc * 2 + 1;
            _  -> error(invalid_character)
        end
    end, 0, Str).

File: archived/src/go/ch01/1.1/l0/bin_to_dec.go
package main

import "errors"

// BinToDec parses a binary string like "1011" into a non-negative integer.
func BinToDec(s string) (uint, error) {
	if len(s) == 0 {
		return 0, errors.New("empty string")
	}
	var v uint
	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '0':
			v = v << 1
		case '1':
			v = (v << 1) | 1
		default:
			return 0, errors.New("invalid character (expected '0' or '1')")
		}
	}
	return v, nil
}
